File: QuickSort.py
def quick_sort(arr:[],low:int,high:int):
    '''
    Here we take a pivot element and find the correct position of it.
    Once we find this index, the elements on left of it will be lower than pivot and 
    the elements on the right would be greater than the pivot.

    We continue the same for left sub array and right sub array using recursion
    '''
    if low<high:
        index = partition(arr,low,high)
        quick_sort(arr,low,index-1)
        quick_sort(arr,index+1,high)
def partition(arr:[],low:int,high:int)->int:
    pivot = arr[low]
    i,j=low,high
    while i<j:
        while arr[i]<=pivot and i<=high-1: # here the loop runs i gets incremented until u find the first value which is GREATER than pivot
            # and here i<=high-1 because if we we put i<=high we would be checking i==4 and then we would be incrementing i to length + 1 and it would be out of bounds
            i+=1
        while arr[j]>=pivot and j>=low+1: # here the loop runs j gets decremented until u find the first value which is SMALLER than pivot
            j-=1 # here j>=low+1 because if we put j>=low then j would be going to j=0 and then decrement j by -1 which is out of bound
        if i<j:
            arr[i],arr[j]=arr[j],arr[i]
    arr[j],arr[low]=arr[low],arr[j]
    return j
if __name__ == "__main__":
    lst =[4,3,2,1,5,5]
    quick_sort(lst,0,len(lst)-1)
    print(lst)

File: MyPrepJava/src/QuickSort.java
import java.util.ArrayList;
import java.util.List;

public class QuickSort {


    void quicksort(List<Integer> list, int low, int high){
        if(low<high){
            int index = partition(list, low,high);
            quicksort(list,low,index-1);
            quicksort(list,index+1,high);
        }
    }
    int partition(List<Integer> list, int low, int high){
        int pivot = list.get(low);
        int i=low, j=high;
        while(i<j){
            while(list.get(i)<=pivot && i<=high-1){
                i++;
            }
            while(list.get(j) >pivot && j>=low+1){
                j--;
            }
            if(i<j){
                int temp=list.get(j);
                list.set(j, list.get(i));
                list.set(i,temp);
            }
        }
        int temp=list.get(j);
        list.set(j, list.get(low));
        list.set(low,temp);
        return j;
    }
    public static void main(String[] args){
        QuickSort q = new QuickSort();
        List<Integer> list = new ArrayList<>();
        list.add(3);
        list.add(5); list.add(1); list.add(9);  list.add(0);
        q.quicksort(list,0,list.size()-1);
        System.out.println(list);
    }
}
